### PR TITLE
Port runtime's Parallel.ForEachAsync so we can use it on NetFx.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:467]: pinged Stephen as i'm not sure what to do here.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:47]: for now, just disabling the arg checks internally.  we rarely ever check this with our ienumerable helpers.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:96]: ported these two helpers over as well.  they're normally internal properties on ParallelOptions, so we don't have direct access to them anymore.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:388]: i think it should be safe to drop IThreadPoolWorkItem.  Though i wouldn't mind @stephentoub  confirming.

Switched to the generic TCS as that's all that netfx has.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:439]: i think this should be safe.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:564]: i think this should be ok.  checking with @stephentoub .

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (stephentoub) [reply src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:382]: > i think it should be safe to drop IThreadPoolWorkItem. Though i wouldn't mind @stephentoub confirming.

On .NET Core that serves as an optimization to avoid allocating each time it's queued. It's functionally fine to drop it.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (stephentoub) [reply src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:433]: Yes, the use of Unsafe is just an optimization to avoid unnecessarily capturing the current ExecutionContext (which is unfortunately more expensive on netfx).

**Comment by CyrusNajmabadi** [comment]
@ToddGrun  this is ready for review.  Will be holding off from merging this until i get OK from stephen on the changes i made.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs]: have applied these suggestions.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:382]: sgtm.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:438]: sgtm.

**Comment by CyrusNajmabadi** [comment]
Ok.  Looks like Stephen is good.  @ToddGrun  this is ready for review.

File: src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs:14]: to be very clear.  this is effectively the runtime's code.  just tweaked as *little* as possible to get working on netfx.

i do *not* want to change it beyond that, as that will be much harder to keep in sync and validate against the netcore impl.

**Comment by ToddGrun** [comment]
Did a cursory review, I'm going to trust Stephen and Cyrus on this one.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski For review when you get back.

**Pull request merged**
